aten/src/ATen/native/native_functions.yaml

```diff
@@ -1233,13 +1233,25 @@
     CompositeExplicitAutograd: copysign_out
   tags: pointwise
 
+# This function is only used for testing COW (Copy-on-Write) behavior.
+# If `torch.get_future_lazy_clone() == False`, the output is a view of the input, but
+# COW behavior is simulated. Otherwise, the output is a semantic copy of the
+# input, but it is a COW tensor, and any aliases of the tensor are also
+# converted to COW tensors.
 - func: _lazy_clone(Tensor self) -> Tensor
-  # Like clone, but the copy takes place lazily, only if either the
-  # input or the output are written.
   variants: function, method
   dispatch:
     CompositeExplicitAutograd: _lazy_clone
 
+# This function applies COW (if `torch.get_future_lazy_clone() == True`) or
+# simulated COW behavior (if `torch.get_future_lazy_clone() == False`) to the
+# input. It should be called on the output of any operator that currently
+# conditionally returns either a view or a copy of the input, in cases where a
+# view is returned, in order to deprecate the conditional view and enable future
+# lazy clone behavior.
+- func: _apply_cow_(Tensor(a!) self) -> Tensor(a!)
```

Comment:
@ezyang , I'm having a lot of trouble getting `_apply_cow_` registered properly so that it passes CI for functorch. Originally, I tried not even registering it, since there's no need to call it from Python. But that broke `reshape`'s vmap support, so I figured I needed to register `_apply_cow_`. The current way that it's registered passes some of the tests that were failing before, but it still fails some of the tests. I've been trying various different ways to register it, and none of them have worked in all cases yet. So I guess two main questions I'm trying to answer are:

* Do I need to register `_apply_cow_` in order for operators that call it to support functorch?
* How do I correctly register an in-place function that doesn't have out-place variants, so that other operators that call the function can support functorch?

Reply:
@zou3519, wondering if you might have some advice about this?

Reply:
Looks like I might have to do something similar to `resize_`. https://github.com/pytorch/pytorch/blob/e76b28c7658561de1bcaff9b5af7156e1d3f3880/aten/src/ATen/FunctionalTensorWrapper.cpp#L311

Reply:
Actually, I realized that `Tensor.set_` is so similar to `_apply_cow_` that I should be able to copy from it. I'll try that out.

Reply:
Nevermind, that doesn't work exactly because I need `_apply_cow_` to support autograd, which neither `set_` nor `resize_` do.

Reply:
Hmm. cc @zou3519 but I think we need to figure out how CoW is supposed to work with functorch :think:

Reply:
Do you have some test cases I could work my head through? Is this just calling vmap over reshape()?

Reply:
So, one thing that pops out to me, is that function is registered as mutating. Is it possible to do this function purely functionally? Perhaps, maybe you do not even want a dispatched operator in native_functions.yaml at all?

Reply:
> Do you have some test cases I could work my head through?

One of the tests that fails is `python test/functorch/test_vmap.py -k test_unsafe_view`, which raises the error "NotImplementedError: Cannot access storage of BatchedTensorImpl". I'm not 100% sure of all that happens under the hood in that test, but it seems that a vmapped call to reshape must be happening, since `_apply_cow_` is given a batched tensor. And `_apply_cow_` needs to manipulate the storage of the tensor given to it, which is not allowed for batched tensors.

> Is it possible to do this function purely functionally?

That's a good point, it should be possible to just use `_lazy_clone` here. Trying that out, it does fix the test I mentioned above (`test_unsafe_view`), but it causes some new test failures. I'll see if I can fix those.

> Perhaps, maybe you do not even want a dispatched operator in native_functions.yaml at all?

Right, I don't necessarily need a dispatched operator. But when `reshape_` is called with a batched tensor, the function that applies the COW/COWSim context to the output (`_apply_cow_` or `_lazy_clone`, whichever one we end up using) needs to also support batched tensors. If there's a way to support that without registering the function, then that could be the way to go

Reply:
It seems more feasible to ensure lazy_clone works correctly with batched tensors

Reply:
Ok I've removed `_apply_cow_` in favor of just calling `_lazy_clone` in `reshape`, but I'm still stuck.

One of the failures I get now is:

```bash
$ python test/functorch/test_vmap.py -k test_reshape
...
======================================================================
FAIL: test_reshape (__main__.TestVmapOperators)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/kurtamohler/develop/pytorch-1/test/functorch/test_vmap.py", line 1381, in wrapper
    method(*args, **kwargs)
  File "/home/kurtamohler/develop/pytorch-1/torch/testing/_internal/common_utils.py", line 2728, in wrapper
    method(*args, **kwargs)
  File "/home/kurtamohler/develop/pytorch-1/test/functorch/test_vmap.py", line 2853, in test_reshape
    test(op, (torch.rand(B0, 2 * 5), [2, 5]), in_dims=(0, None), check_view=True)
  File "/home/kurtamohler/develop/pytorch-1/test/functorch/test_vmap.py", line 1423, in _vmap_test
    return _vmap_test(self, *args, **kwargs)
  File "/home/kurtamohler/develop/pytorch-1/test/functorch/test_vmap.py", line 1309, in _vmap_test
    self.assertTrue(
AssertionError: False is not true : result was not a view of the first input!
```

I tried changing `_lazy_clone` to just return the input, unmodified, and I still get that failure. So evidently, the implementation of `_lazy_clone` is not the issue.

When I change `reshape` to no longer call `_lazy_clone`, the failure goes away. So I assume the issue is with how functorch interacts with `_lazy_clone` or how `_lazy_clone` is registered.

I'm guessing that I need to change something to tell functorch that `_lazy_clone` can return a view of the input. I tried adding the `(a)` alias annotation to it, but then I get this error during build:
```bash
 File "/home/kurtamohler/develop/pytorch-1/torchgen/gen.py", line 1416, in get_grouped_by_view_native_functions
    return list(concatMap(maybe_create_view_group, grouped_by_views.values()))
  File "/home/kurtamohler/develop/pytorch-1/torchgen/utils.py", line 86, in concatMap
    yield from func(x)
  File "/home/kurtamohler/develop/pytorch-1/torchgen/gen.py", line 1385, in maybe_create_view_group
    NativeFunctionsViewGroup(
  File "<string>", line 6, in __init__
  File "/home/kurtamohler/develop/pytorch-1/torchgen/model.py", line 2659, in __post_init__
    assert not gets_generated_view_copy(self.view), (
AssertionError: _lazy_clone appears to be a new operator that aliases its inputs. The codegen expects you to add a corresponding operator to native_functions.yaml: _lazy_clone_copy. See Note [view_copy NativeFunctions] for details.
```

So I tried adding the operator `_lazy_clone_copy`, copying what I see for other `*_copy` operators:

```
- func: _lazy_clone_copy(Tensor self) -> Tensor
  variants: function
  tags: view_copy
  dispatch:
    CompositeExplicitAutogradNonFunctional: _lazy_clone_copy
```

That builds, but when I try to import PyTorch, I get:

```bash
ImportError: /home/kurtamohler/develop/pytorch-1/torch/lib/libtorch_cpu.so: undefined symbol: _ZN2at17functionalization18FunctionalInverses19_lazy_clone_inverseERKNS_6TensorES4_NS0_17InverseReturnModeE
```

I'm not sure what `_lazy_clone_inverse` is supposed to be, but I'll try to add it.

So I still have a path to go down, but I just want to ask if I'm taking the right steps, or if anyone has any suggestions.

Reply:
`_lazy_clone` is morally not a view operation (it is morally a clone), so it should not get the alias annotation. Let me check out your branch and see what's up

Reply:
Just to be clear, with the backward compatible default behavior, `_lazy_clone` just creates a view. The view has a new StorageImpl, but from the front end, it is supposed to be treated as a view. With future BC-breaking behavior turned on, you're right, it is a clone

Reply:
@kurtamohler I think you just need to keep working the errors. If you grep inverse you'll find all the preexisting ones, should be straightforward. also cc @bdhirsh

Reply:
I got the aliasing annotations to work, and some of the tests that were failing before are now passing. However, some are still failing, including these two:

```bash
$ python test/functorch/test_vmap.py -k test_reshape
...
  File "/home/kurtamohler/develop/pytorch-1/test/functorch/test_vmap.py", line 1423, in _vmap_test
    return _vmap_test(self, *args, **kwargs)
  File "/home/kurtamohler/develop/pytorch-1/test/functorch/test_vmap.py", line 1309, in _vmap_test
    self.assertTrue(
AssertionError: False is not true : result was not a view of the first input!
```

```bash
$ python test/test_ops.py -k test_python_ref__refs_reshape_cpu_float32
...
  File "/home/kurtamohler/develop/pytorch-1/test/test_ops.py", line 422, in _ref_test_helper
    self.assertEqual(a._is_view(), b._is_view(), msg)
  File "/home/kurtamohler/develop/pytorch-1/torch/testing/_internal/common_utils.py", line 3620, in assertEqual
    raise error_metas.pop()[0].to_error(
AssertionError: Booleans mismatch: True is not False
The torch implementation does not return a view, while the reference does
```

The second one (and maybe the first one as well) evidently happens because the output of `reshape` gives `_is_view() == False`:

```bash
$ python -c 'import torch; a = torch.randn(10); b = a.reshape(a.size()); print(b._is_view())'
False
```

And the reason for that is presumably because the output of `_lazy_clone` has `_is_view() == False` as well:

```bash
$ python -c 'import torch; a = torch.randn(10); b = a._lazy_clone(); print(b._is_view())'
False
```

If I try changing `_lazy_clone` to just return the input without doing anything else, both of the above tests pass, and the output of `torch.reshape` has `_is_view() == True`. But even so, the output of `torch._lazy_clone` still has `_is_view() == False`, because it just returns the input directly, not a view.

I also tried changing `_lazy_clone` to just return a normal view of the input (`return self.view_symint(self.sym_sizes())`) without instrumenting it with COW. With this change, the tests still fail, and the outputs of both `torch.reshape` and `torch._lazy_clone` give `_is_view() == False`. I'm not sure why. Apparently there must be something outside of the implementation of `_lazy_clone` that I still need to fix to make its output have `_is_view() == True`.

Also, just to make sure, I tried removing the `_lazy_clone` calls in `reshape` and call `view_symint` instead. The `reshape` tests pass as I expected.

I've spent a little time trying to figure out how to make `_lazy_clone`'s output give `_is_view() == True`, but I haven't figured it out yet. I'll keep looking, but if someone else knows, please let me know.

Reply:
@albanD anything pop out to you here? I feel like if I have an alias annotation in my schema, the generated autograd function in C++ ought to setup ViewMeta.

@kurtamohler I suggest popping open VariableTypesEverything.cpp and looking at lazy_clone's generated code. Post it here.

Reply:
Here's what's in VariableTypeEverything.cpp:

<details><summary>Click to expand</summary>

```cpp
at::Tensor _lazy_clone(c10::DispatchKeySet ks, const at::Tensor & self) {
  auto& self_ = unpack(self, "self", 0);
  [[maybe_unused]] auto _any_requires_grad = compute_requires_grad( self );
  
  [[maybe_unused]] auto _any_has_forward_grad_result = (isFwGradDefined(self));
  std::shared_ptr<LazyCloneBackward0> grad_fn;
  if (_any_requires_grad) {
    grad_fn = std::shared_ptr<LazyCloneBackward0>(new LazyCloneBackward0(), deleteNode);
    grad_fn->set_next_edges(collect_next_edges( self ));
  }
  #ifndef NDEBUG
  c10::optional<Storage> self__storage_saved =
    self_.has_storage() ? c10::optional<Storage>(self_.storage()) : c10::nullopt;
  c10::intrusive_ptr<TensorImpl> self__impl_saved;
  if (self_.defined()) self__impl_saved = self_.getIntrusivePtr();
  #endif
  auto _tmp = ([&]() {
    at::AutoDispatchBelowADInplaceOrView guard;
    return at::redispatch::_lazy_clone(ks & c10::after_autograd_keyset, self_);
  })();
  auto result = std::move(_tmp);
  #ifndef NDEBUG
  if (self__storage_saved.has_value() &&
      !at::impl::dispatch_mode_enabled() &&
      !at::impl::tensor_has_dispatch(self_) &&
      !at::impl::tensor_has_dispatch(self_))
    TORCH_INTERNAL_ASSERT(self__storage_saved.value().is_alias_of(self_.storage()));
  if (self__impl_saved && !at::impl::dispatch_mode_enabled() && !at::impl::tensor_has_dispatch(self_))
    TORCH_INTERNAL_ASSERT(self__impl_saved == self_.getIntrusivePtr());
  if (result.has_storage() && !at::impl::dispatch_mode_enabled() && !at::impl::tensor_has_dispatch(result)) {
    TORCH_INTERNAL_ASSERT(result.storage().use_count() == 1, "function: _lazy_clone");
  }
  if (!at::impl::dispatch_mode_enabled() && !at::impl::tensor_has_dispatch(result))
    TORCH_INTERNAL_ASSERT(result.use_count() <= 1, "function: _lazy_clone");
  #endif
  if (grad_fn) {
      set_history(flatten_tensor_args( result ), grad_fn);
  }
  ::std::optional<at::Tensor> result_new_fw_grad_opt = c10::nullopt;
  if (_any_has_forward_grad_result && (result.defined())) {
      auto self_t_raw = toNonOptFwGrad(self);
      auto self_tensor = toNonOptTensor(self);
      auto self_t = (self_t_raw.defined() || !self_tensor.defined())
        ? self_t_raw : at::_efficientzerotensor_symint(self_tensor.sym_sizes(), self_tensor.options());
      result_new_fw_grad_opt = at::_lazy_clone(self_t);
  }
  if (result_new_fw_grad_opt.has_value() && result_new_fw_grad_opt.value().defined() && result.defined()) {
    // The hardcoded 0 here will need to be updated once we support multiple levels.
    result._set_fw_grad(result_new_fw_grad_opt.value(), /* level */ 0, /* is_inplace_op */ false);
  }
  return result;
}
```

</details>

Reply:
I decided one way to figure this out is to learn how `Tensor._is_view()` works. After reading through the related code I realized that I just had to add `_lazy_clone` to the `VIEW_FUNCTIONS` dict in `tools/autograd/gen_inplace_or_view_type.py`. That fixes the errors because it adds `_lazy_clone` to `torch/csrc/autograd/generated/ADInplaceOrViewTypeEverything.cpp`, and it wraps the result with a call to `torch::autograd::as_view`, which properly sets the tensor's `ViewMeta` to an instance of `DifferentiableViewMeta`

Reply:
Although, I am going to have to make `Tensor._is_view()` return false when future behavior is enabled

Reply:
reshape is not in VIEW_FUNCTIONS because reshape is a composite. When you reroute reshape to go to lazy_clone, you will automatically start producing non view, which should be good.

Reply:
Also, glad you figured it out! We should probably make it harder to forget to add functions to that list.

Reply:
At the moment, when future behavior is enabled, `_lazy_clone`'s output incorrectly has `_is_view() == True`, even after materializing:

```python
>>> import torch
>>> torch.set_future_lazy_clone(True)
>>> a = torch.randn(10)
>>> b = a._lazy_clone()
>>> b._is_view()
True
>>> b.data_ptr()  # causes COW materialize
94352947356608
>>> a.data_ptr()
94352860738944
>>> b._is_view()
True
```

Both of those `b._is_view()` calls should be False, so I'll need to somehow avoid the `torch::autograd::as_view` call in `torch/csrc/autograd/generated/ADInplaceOrViewTypeEverything.cpp`, in the case when the returned tensor is a real COW. Maybe there are more than two ways to do it, but the ways I can think of are: (1) write a custom function for `ADInplaceOrViewTypeEverything` or (2) add a `_lazy_clone_alias` op for the simulated COW case, which would do the `torch::autograd::as_view` call, and disable the `as_view` call for `_lazy_clone` again.

Reply:
Oh oops. The problem is you can't have an operator whose schema simultaneously aliases and not aliases at the same time. Maybe your other idea is good: have a separate op for future and old behavior, and then have a composite do the switching between them. Only the old behavior has the alias annotation and special func